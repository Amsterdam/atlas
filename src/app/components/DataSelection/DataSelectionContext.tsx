import { useMemo, useState } from 'react'
import type { Dispatch, FunctionComponent, SetStateAction } from 'react'
import type { DataSelectionFilters } from '../../pages/MapPage/query-params'
import { dataSelectionFiltersParam, polygonParam } from '../../pages/MapPage/query-params'
import createNamedContext from '../../utils/createNamedContext'
import useParam from '../../utils/useParam'
import useRequiredContext from '../../utils/useRequiredContext'
import type { ActiveFilter, AvailableFilter } from './types'
import useLegacyDataselectionConfig from './useLegacyDataselectionConfig'

type Action<T extends keyof DataSelectionContextProps> = Dispatch<
  SetStateAction<DataSelectionContextProps[T]>
>

export interface DataSelectionContextProps {
  availableFilters: AvailableFilter[]
  addFilter: (filter: DataSelectionFilters) => void
  removeFilter: (key: string) => void
  totalResults: number
  activeFilters: ActiveFilter[]
  setTotalResults: Action<'totalResults'>
  setAvailableFilters: Action<'availableFilters'>
  distanceText?: string
  drawToolLocked: boolean
  setDrawToolLocked: Action<'drawToolLocked'>
  setDistanceText: Action<'distanceText'>
}

const DataSelectionContext = createNamedContext<DataSelectionContextProps | null>(
  'DataSelection',
  null,
)

const DataSelectionProvider: FunctionComponent = ({ children }) => {
  const [activeParamFilters, setActiveParamFilters] = useParam(dataSelectionFiltersParam)
  const [polygon] = useParam(polygonParam)
  const [distanceText, setDistanceText] = useState<string>()
  const [drawToolLocked, setDrawToolLocked] = useState(false)
  const [availableFilters, setAvailableFilters] = useState<
    DataSelectionContextProps['availableFilters']
  >([])
  const [totalResults, setTotalResults] = useState(0)
  const { currentDatasetConfig } = useLegacyDataselectionConfig()

  const addFilter = (filter: DataSelectionFilters) => {
    setActiveParamFilters({
      ...activeParamFilters,
      ...filter,
    })
  }
  const removeFilter = (key: string) => {
    // @ts-ignore
    const { [key]: omit, ...rest } = activeParamFilters
    setActiveParamFilters(Object.keys(rest).length ? rest : null)
  }

  const shape = useMemo(
    () =>
      polygon?.polygon?.length
        ? JSON.stringify(polygon.polygon.map(({ lat, lng }) => [lng, lat]))
        : null,
    [polygon],
  )

  const activeFilters = useMemo(() => {
    const filters =
      activeParamFilters && currentDatasetConfig
        ? Object.entries(activeParamFilters).reduce<ActiveFilter[]>((acc, [key, val]) => {
            const label = currentDatasetConfig.FILTERS.find(({ slug }) => slug === key)?.label
            if (label) {
              let value = val
              if (key === 'sbi_code') {
                value = val.replace(/['[\]]/g, '')
              }
              acc.push({ key, value, label: `${label}: ${value}` })
            }
            return acc
          }, [])
        : []

    if (shape) {
      filters.push({
        key: 'shape',
        value: shape,
        label: 'Ingetekende vorm op kaart',
      })
    }

    return filters
  }, [currentDatasetConfig, activeParamFilters, shape])

  return (
    <DataSelectionContext.Provider
      value={{
        // Only provide the token if the user does not have an authenticated session.
        availableFilters,
        activeFilters,
        addFilter,
        removeFilter,
        totalResults,
        setTotalResults,
        setAvailableFilters,
        distanceText,
        setDistanceText,
        drawToolLocked,
        setDrawToolLocked,
      }}
    >
      {children}
    </DataSelectionContext.Provider>
  )
}

export { DataSelectionProvider }

export function useDataSelection() {
  return useRequiredContext(DataSelectionContext)
}
