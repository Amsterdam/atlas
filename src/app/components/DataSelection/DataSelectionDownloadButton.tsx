import { Button, themeSpacing } from '@amsterdam/asc-ui'
import { Download } from '@amsterdam/asc-assets'
import { useMatomo } from '@datapunt/matomo-tracker-react'
import styled from 'styled-components'
import type { FunctionComponent } from 'react'
import environment from '../../../environment'
import DATA_SELECTION_CONFIG from '../../../shared/services/data-selection/data-selection-config'
import { encodeQueryParams } from '../../../shared/services/query-string-parser/query-string-parser'
import { getAccessToken } from '../../../shared/services/auth/auth'
import type { ActiveFilter } from './types'
import { DatasetType } from './types'

export interface DataSelectionDownloadButtonProps {
  dataset: DatasetType
  activeFilters: ActiveFilter[]
}

const StyledButton = styled(Button)`
  margin-left: ${themeSpacing(4)};
`

const DataSelectionDownloadButton: FunctionComponent<DataSelectionDownloadButtonProps> = ({
  dataset,
  activeFilters,
}) => {
  const { trackEvent } = useMatomo()
  const filterParams = []
  let url = `${environment.API_ROOT}${
    DATA_SELECTION_CONFIG.datasets[dataset].ENDPOINT_EXPORT ?? ''
  }`

  DATA_SELECTION_CONFIG.datasets[dataset].FILTERS.forEach((filter: { slug: string }) => {
    const activeFilter = activeFilters.find(({ key }) => key === filter.slug)
    if (activeFilter) {
      filterParams.push(`${filter.slug}=${window.encodeURIComponent(activeFilter.value)}`)
    }
  })

  const shape = activeFilters.find(({ key }) => key === 'shape')
  if (shape) {
    filterParams.push(`shape=${shape.value}`)
  }

  if (dataset === DatasetType.Hr) {
    filterParams.push(DATA_SELECTION_CONFIG.datasets[dataset].ENDPOINT_EXPORT_PARAM)
  }

  if (filterParams.length) {
    url += `?${filterParams.join('&')}`
  }

  const token = getAccessToken()

  const params = {}
  if (token) {
    // @ts-ignore
    params.access_token = token
  }

  const queryStart = url.indexOf('?') !== -1 ? '&' : '?'
  const paramString = encodeQueryParams(params)
  const queryString = paramString ? queryStart + paramString : ''
  const downloadUrl = `${url}${queryString}`

  return (
    <StyledButton
      forwardedAs="a"
      variant="primary"
      href={downloadUrl}
      iconSize={21}
      iconLeft={<Download />}
      onClick={() => {
        trackEvent({
          category: 'Download-tabel',
          action: `dataselectie-download-${DATA_SELECTION_CONFIG.datasets[
            dataset
          ].TITLE.toLowerCase()}`,
        })
      }}
      title="Downloaden als kommagescheiden bestand"
    >
      Downloaden
    </StyledButton>
  )
}

export default DataSelectionDownloadButton
