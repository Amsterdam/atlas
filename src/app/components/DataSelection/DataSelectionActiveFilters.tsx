import { useDispatch } from 'react-redux'
import { FilterTag } from '@amsterdam/asc-ui'
import styled from 'styled-components'
import { removeGeometryFilter } from '../../../shared/ducks/data-selection/actions'
import { useDataSelection } from './DataSelectionContext'
import useParam from '../../utils/useParam'
import { polygonParam } from '../../pages/MapPage/query-params'

const List = styled.ul`
  display: flex;
  flex-wrap: wrap;
  list-style: none;
`

const DataSelectionActiveFilters = () => {
  const dispatch = useDispatch()
  const { removeFilter, activeFilters } = useDataSelection()
  const [, setPolygon] = useParam(polygonParam)
  return activeFilters.length ? (
    <List data-testid="activeFilters">
      {activeFilters.map(({ key, label, value }) => (
        <li key={key}>
          <FilterTag
            type="button"
            aria-label={`Filter verwijderen: ${label}: ${value}`}
            onClick={() => {
              if (key === 'shape') {
                setPolygon(null)
                dispatch(removeGeometryFilter())
              } else {
                removeFilter(key)
              }
            }}
            data-testid="activeFiltersItem"
          >
            {label}
          </FilterTag>
        </li>
      ))}
    </List>
  ) : null
}

export default DataSelectionActiveFilters
