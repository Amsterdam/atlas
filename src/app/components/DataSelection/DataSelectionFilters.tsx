import { useMemo, useState } from 'react'
import { Button, Heading, Link, themeColor, themeSpacing } from '@amsterdam/asc-ui'
import styled from 'styled-components'
import { Enlarge, Minimise } from '@amsterdam/asc-assets'
import type { FunctionComponent } from 'react'
import { DEFAULT_LOCALE } from '../../../shared/config/locale.config'
import Metadata from '../../../shared/assets/icons/metadata.svg'
import { useDataSelection } from './DataSelectionContext'
import type { AvailableFilter } from './types'
import useLegacyDataselectionConfig from './useLegacyDataselectionConfig'

const StyledHeading = styled(Heading)`
  display: flex;
  justify-content: space-between;
  color: ${themeColor('tint', 'level7')};
  margin-bottom: ${themeSpacing(2)};
`

const InfoButton = styled(Button)`
  border: 1px solid ${themeColor('tint', 'level4')};
`

const StyledLink = styled(Link)`
  background-color: transparent;
  text-align: left;
`
const ShowMoreButton = styled(Button)`
  margin-top: ${themeSpacing(2)};
`

// Note, this component has been migrated from legacy angularJS code
// Todo: this can be removed when we implement the new interactive table
const DataSelectionFilters: FunctionComponent = () => {
  const [expandedFilters, setExpandedFilters] = useState<string[]>([])
  const { addFilter, activeFilters, availableFilters } = useDataSelection()
  const { currentDatasetConfig } = useLegacyDataselectionConfig()
  const filteredAvailableFilters = useMemo(() => {
    if (!activeFilters.length) {
      return []
    }
    return availableFilters.filter(({ slug }) => activeFilters.find(({ key }) => key === slug))
  }, [activeFilters, availableFilters])

  const showMoreThreshold = 10

  const isFilterOptionActive = (filterSlug: string, id: string, label: string) => {
    return filteredAvailableFilters.find(({ slug }) => slug === label || slug === id)
  }

  const hasInactiveFilterOptions = (filter: AvailableFilter) => {
    return (
      !filter.options.some((option) =>
        isFilterOptionActive(filter.slug, option.id, option.label),
      ) ||
      (!filter.slug.startsWith('sbi') && filter.options.length)
    )
  }

  const isExpandedFilter = (filterSlug: string) => {
    return expandedFilters.indexOf(filterSlug) !== -1
  }

  const showExpandButton = (filterSlug: string) => {
    return (
      !isExpandedFilter(filterSlug) && getAvailableOptions(filterSlug).length > showMoreThreshold
    )
  }

  const nrHiddenOptions = (filter: AvailableFilter) => {
    return filter.numberOfOptions - filter.options.length
  }

  const expandFilter = (filterSlug: string) => {
    setExpandedFilters((currentExpandedFilters) => [...currentExpandedFilters, filterSlug])
  }

  const implodeFilter = (filterSlug: string) => {
    setExpandedFilters((currentExpandedFilters) => [
      ...currentExpandedFilters.filter((val) => val !== filterSlug),
    ])
  }

  const canExpandImplode = (filterSlug: string) => {
    return getAvailableOptions(filterSlug).length > showMoreThreshold
  }

  function getAvailableOptions(filterSlug: string) {
    return getAvailableFilters(filterSlug)[0].options
  }

  function getAvailableFilters(filterSlug: string) {
    return availableFilters.filter((filter) => filter.slug === filterSlug)
  }

  if (!currentDatasetConfig) {
    return null
  }

  const showOptionCounts = currentDatasetConfig.SHOW_FILTER_OPTION_COUNTS
  const activeFilterSlugs = filteredAvailableFilters.map(({ slug }) => slug)
  return (
    <div
      className="qa-available-filters c-data-selection-available-filters"
      data-testid="dataSelectionAvailableFilters"
    >
      <>
        {availableFilters.map(
          (filter) =>
            !activeFilterSlugs.includes(filter.slug) &&
            hasInactiveFilterOptions(filter) && (
              <div className="c-data-selection-available-filters__category" key={filter.slug}>
                <StyledHeading as="h3">
                  <span>{filter.label}</span>
                  {filter.info_url && (
                    <InfoButton
                      type="button"
                      variant="blank"
                      data-testid="detail-infobox"
                      forwardedAs={Link}
                      size={28}
                      iconSize={18}
                      icon={<Metadata />}
                      href={filter.info_url}
                      rel="noreferrer noopener"
                      target="_blank"
                    >
                      Uitleg tonen
                    </InfoButton>
                  )}
                </StyledHeading>

                <ul>
                  {filter.options
                    .slice(
                      0,
                      isExpandedFilter(filter.slug) ? filter.options.length : showMoreThreshold,
                    )
                    .map((option) => (
                      <li key={option.id} className="c-data-selection-available-filters__item">
                        <StyledLink
                          inList
                          type="button"
                          forwardedAs="button"
                          onClick={() => {
                            addFilter({
                              [filter.slug]: option.id,
                            })
                          }}
                        >
                          <span className="c-data-selection-available-filters__item-label">
                            <span className="qa-option-label">{option.label || '(geen)'}</span>
                            {showOptionCounts && <span>({option.count})</span>}
                          </span>
                        </StyledLink>
                      </li>
                    ))}
                </ul>

                {canExpandImplode(filter.slug) && showExpandButton(filter.slug) && (
                  <div className="c-data-selection-available-filters__show-more">
                    <ShowMoreButton
                      type="button"
                      onClick={() => expandFilter(filter.slug)}
                      className="qa-show-more-button"
                      variant="textButton"
                      iconSize={14}
                      iconLeft={<Enlarge />}
                    >
                      Toon meer
                    </ShowMoreButton>
                  </div>
                )}

                {canExpandImplode(filter.slug) && isExpandedFilter(filter.slug) && (
                  <div className="c-data-selection-available-filters__show-more">
                    {nrHiddenOptions(filter) > 0 && (
                      <div className="c-data-selection-available-filters__hidden-options qa-hidden-options">
                        ...nog {nrHiddenOptions(filter).toLocaleString(DEFAULT_LOCALE)} waarden
                      </div>
                    )}
                    <ShowMoreButton
                      type="button"
                      className="qa-show-more-button"
                      onClick={() => implodeFilter(filter.slug)}
                      variant="textButton"
                      iconSize={14}
                      iconLeft={<Minimise />}
                    >
                      Toon minder
                    </ShowMoreButton>
                  </div>
                )}
              </div>
            ),
        )}
      </>
    </div>
  )
}

export default DataSelectionFilters
