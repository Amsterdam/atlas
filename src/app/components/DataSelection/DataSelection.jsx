import React from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import classNames from 'classnames';
import { bindActionCreators } from 'redux';
import {
getDataSelection,
getDataSelectionResult,
setPage as setDatasetPage,
VIEWS
} from '../../../shared/ducks/data-selection/data-selection';
import DATA_SELECTION_CONFIG from '../../../shared/services/data-selection/data-selection-config';
import { getUser } from '../../../shared/ducks/user/user';
import NotAuthorizedPanel from '../PanelMessages/NotAuthorizedMessage';
import LoadingIndicator from '../../../shared/components/loading-indicator/LoadingIndicator';
import { getActiveFilters } from '../../../shared/ducks/filters/filters';
import DataSelectionActiveFilters from '../../containers/DataSelectionActiveFiltersContainer';
import MaxPageMessage from '../PanelMessages/MaxPageMessage';
import AngularWrapper from '../../../react-angular/AngularWrapper';

const DataSelection = ({
  view,
  isLoading,
  dataset,
  activeFilters,
  user,
  setPage,
  authError,
  results: {
    numberOfRecords,
    filters: availableFilters,
    numberOfPages,
    data
  },
  page: currentPage
}) => {
  if (isLoading || (!numberOfRecords && !authError)) {
    return <LoadingIndicator />;
  }

  // Local state
  const showHeader = (view === VIEWS.LIST || !isLoading);
  const showFilters = (view !== VIEWS.LIST) && numberOfRecords > 0;
  const { MAX_AVAILABLE_PAGES, MAX_NUMBER_OF_CLUSTERED_MARKERS } =
    DATA_SELECTION_CONFIG.datasets[dataset];

  const showMessageMaxPages = MAX_AVAILABLE_PAGES && currentPage > MAX_AVAILABLE_PAGES;
  const showMessageClusteredMarkers =
    (view === VIEWS.LIST) && numberOfRecords > MAX_NUMBER_OF_CLUSTERED_MARKERS;

  const widthClass = classNames({
    'u-col-sm--12': !showFilters,
    'u-col-sm--9': showFilters
  });

  return (
    <div className="c-data-selection">
      <div className="qa-data-selection-content">
        <AngularWrapper
          moduleName={'dpDataSelectionHeaderWrapper'}
          component="dpDataSelectionHeader"
          dependencies={['atlas']}
          bindings={{
            dataset,
            availableFilters,
            filters: activeFilters,
            numberOfRecords,
            showHeader,
            user,
            view
          }}
        />

        <DataSelectionActiveFilters />

        {(!authError) && (
          <div className="u-grid qa-data-grid">
            <div className="u-row">
              {showFilters && (
                <div className="u-col-sm--3 c-data-selection__available-filters">
                  {(dataset === 'hr') && (
                    <AngularWrapper
                      moduleName="dpSbiFilterWrapper"
                      component="dpSbiFilter"
                      dependencies={['atlas']}
                      bindings={{
                        availableFilters,
                        filters: activeFilters
                      }}
                    />
                  )}

                  <AngularWrapper
                    moduleName="dpDataSelectionAvailableFiltersWrapper"
                    component="dpDataSelectionAvailableFilters"
                    dependencies={['atlas']}
                    bindings={{
                      availableFilters,
                      activeFilters
                    }}
                    interpolateBindings={{
                      dataset
                    }}
                  />
                </div>
              )}
              <div className={widthClass}>
                {showMessageMaxPages && (
                  <MaxPageMessage maxAvailablePages={MAX_AVAILABLE_PAGES} />
                )}
                {showMessageClusteredMarkers && (
                  <AngularWrapper
                    moduleName={'dpPanelWrapper'}
                    component="dpPanel"
                    dependencies={['atlas']}
                    bindings={{
                      isPanelVisible: true,
                      canClose: true
                    }}
                    interpolateBindings={{
                      type: 'warning'
                    }}
                  >
                    <div className="qa-message-clustered-markers">
                      <p className="c-panel__paragraph">Deze resultaten worden niet getoond op
                        de
                        kaart, omdat deze niet meer
                        dan {MAX_NUMBER_OF_CLUSTERED_MARKERS} resultaten tegelijk kan
                        weergeven (om technische redenen).</p>
                      <p className="c-panel__paragraph">Tip: Bekijk de lijst resultaten in
                        kleinere delen. Dit kan door een voor een filtercriteria toe te voegen
                        (bijv. de verschillende wijken uit de selectie).</p>
                    </div>
                  </AngularWrapper>
                )}


                <div>
                  {view === VIEWS.TABLE && (
                    <AngularWrapper
                      moduleName={'dpDataSelectionTableWrapper'}
                      component="dpDataSelectionTable"
                      dependencies={['atlas']}
                      bindings={{
                        content: data,
                        dataset
                      }}
                    />
                  )}
                  {view === VIEWS.LIST && (
                    <AngularWrapper
                      moduleName={'dpDataSelectionListWrapper'}
                      component="dpDataSelectionList"
                      dependencies={['atlas']}
                      bindings={{
                        content: data
                      }}
                    />
                  )}
                </div>
                <AngularWrapper
                  moduleName={'dpDataSelectionPaginationWrapper'}
                  component="dpDataSelectionPagination"
                  dependencies={['atlas']}
                  bindings={{
                    currentPage,
                    numberOfPages,
                    setPage
                  }}
                />
              </div>
            </div>
          </div>
        )}

        {authError && (
          <NotAuthorizedPanel />
        )}
      </div>
    </div>
  );
};

DataSelection.defaultProps = {
  results: {}
};

DataSelection.propTypes = {
  view: PropTypes.string.isRequired,
  isLoading: PropTypes.bool.isRequired,
  dataset: PropTypes.string.isRequired,
  activeFilters: PropTypes.shape({}).isRequired,
  user: PropTypes.shape({}).isRequired,
  authError: PropTypes.bool.isRequired,
  page: PropTypes.number.isRequired,
  setPage: PropTypes.func.isRequired,
  results: PropTypes.shape({
    numberOfRecords: PropTypes.number,
    filters: PropTypes.arrayOf(PropTypes.object),
    numberOfPages: PropTypes.number,
    data: PropTypes.object
  })
};

const mapStateToProps = (state) => {
  const { view, isLoading, dataset, authError, page } = getDataSelection(state);
  return ({
    view,
    isLoading,
    dataset,
    authError,
    page,
    activeFilters: getActiveFilters(state),
    results: getDataSelectionResult(state),
    user: getUser(state)
  });
};

const mapDispatchToProps = (dispatch) => bindActionCreators({
  setPage: setDatasetPage
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(DataSelection);
