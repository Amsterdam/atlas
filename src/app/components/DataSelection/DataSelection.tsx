import { Alert, Heading, Paragraph, themeColor, themeSpacing } from '@amsterdam/asc-ui'
import { useSelector } from 'react-redux'
import usePromise, { isPending, isRejected } from '@amsterdam/use-promise'
import styled, { css } from 'styled-components'
import { FunctionComponent } from 'react'
import { getDataSelectionPage } from '../../../shared/ducks/data-selection/selectors'
import DataSelectionActiveFilters from './DataSelectionActiveFilters'
import { ViewMode, viewParam } from '../../pages/MapPage/query-params'
import useParam from '../../utils/useParam'
import LoginLink from '../Links/LoginLink/LoginLink'
import LoadingSpinner from '../LoadingSpinner/LoadingSpinner'
import ShareBar from '../ShareBar/ShareBar'
import DataSelectionList from './DataSelectionList'
import DataSelectionTable from './DataSelectionTable'
import LegacyPagination from './LegacyPagination'
import DataSelectionFilters from './DataSelectionFilters'
import DataSelectionSbiFilters from './DataSelectionSbiFilters'
import useFetchLegacyDataSelectionData from './useFetchLegacyDataSelectionData'
import { useDataSelection } from './DataSelectionContext'
import DataSelectionHeader from './DataSelectionHeader'
import { AuthError } from '../../../shared/services/api/customError'
import useFetchLegacyDataSelectionMarkers from './useFetchLegacyDataSelectionMarkers'

const StyledAlert = styled(Alert)`
  margin-bottom: ${themeSpacing(5)};
`

const StyledContainer = styled.section<{ view: ViewMode }>`
  display: flex;
  flex-direction: column;
  margin: ${themeSpacing(4, 0)};
  padding: ${themeSpacing(0, 4)};
  background-color: ${themeColor('tint', 'level1')};

  ${({ view }) =>
    view === ViewMode.Full &&
    css`
      overflow-x: auto;
    `}
`

const Wrapper = styled.div<{ view: ViewMode }>`
  margin-bottom: ${themeSpacing(4)};
  ${({ view }) =>
    view === ViewMode.Full &&
    css`
      display: grid;
      grid-template-columns: 25% 75%;
      grid-template-rows: 2fr;
      grid-column-gap: ${themeSpacing(2)};
      grid-row-gap: 0;
    `}
`

const Footer = styled.footer`
  grid-column: 2;
`

interface DataSelectionContentProps {
  view: ViewMode
}

const DataSelectionContent: FunctionComponent<DataSelectionContentProps> = ({ view }) => {
  const page = useSelector(getDataSelectionPage)
  const { activeFilters, totalResults } = useDataSelection()
  const { fetchData } = useFetchLegacyDataSelectionData()
  const { fetchMarkers } = useFetchLegacyDataSelectionMarkers()

  const result = usePromise(
    (signal) => fetchData(signal),
    [
      view, // view can change: table or map view AKA Full or Split mode
      page, // Pagination page
      activeFilters, // Enabling or disabling filters
    ],
  )

  usePromise((signal) => fetchMarkers(signal), [result])

  if (isPending(result)) {
    return <LoadingSpinner />
  }

  if (isRejected(result)) {
    if (result.reason instanceof AuthError) {
      return (
        <Alert level="info" dismissible>
          <Paragraph>{result.reason.message}</Paragraph>
          <LoginLink />
        </Alert>
      )
    }

    return (
      <Alert level="warning" dismissible>
        {result.reason.message}
      </Alert>
    )
  }

  const { numberOfPages, data, showSbiFilers, messageMaxPages, messageClusteredMarkers } =
    result.value

  const showFilters = view === ViewMode.Full && totalResults > 0
  return (
    <>
      <DataSelectionActiveFilters />

      {!totalResults && (
        <>
          <Paragraph>Geen resultaten van deze soort</Paragraph>
          <Paragraph>Tip: verwijder een of meer criteria</Paragraph>
        </>
      )}

      <Wrapper view={view}>
        {showFilters && (
          <div>
            {showSbiFilers && <DataSelectionSbiFilters />}
            <DataSelectionFilters />
          </div>
        )}
        <div>
          {messageMaxPages && (
            <StyledAlert level="info" dismissible>
              <Heading forwardedAs="h3">Deze pagina kan niet worden getoond</Heading>
              <Paragraph>{messageMaxPages}</Paragraph>
              <Paragraph>
                Tip: Gebruik de download-knop om alle resultaten te bekijken. Of voeg meer
                filtercriteria toe voor specifiekere resultaten.
              </Paragraph>
            </StyledAlert>
          )}
          {messageClusteredMarkers && (
            <StyledAlert level="info">
              <Paragraph>{messageClusteredMarkers}</Paragraph>
              <Paragraph>
                Tip: Bekijk de lijst resultaten in kleinere delen. Dit kan door een voor een
                filtercriteria toe te voegen (bijv. de verschillende wijken uit de selectie).
              </Paragraph>
            </StyledAlert>
          )}

          {totalResults > 0 ? (
            <div>
              {view === ViewMode.Full && <DataSelectionTable content={data} />}
              {view === ViewMode.Split && <DataSelectionList content={data} />}
            </div>
          ) : null}
        </div>
        <Footer>
          <LegacyPagination currentPage={page} numberOfPages={numberOfPages} />
          {view === ViewMode.Full && <ShareBar />}
        </Footer>
      </Wrapper>
    </>
  )
}

const DataSelection = () => {
  const [view] = useParam(viewParam)
  return (
    <StyledContainer view={view} className="c-data-selection">
      <div className="c-data-selection-content">
        <DataSelectionHeader />
        <DataSelectionContent view={view} />
      </div>
    </StyledContainer>
  )
}

export default DataSelection
