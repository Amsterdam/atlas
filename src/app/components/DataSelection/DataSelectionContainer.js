import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { setPage as setDatasetPage } from '../../../shared/ducks/data-selection/actions';
import { getUser, getUserScopes } from '../../../shared/ducks/user/user';
import { getFilters } from '../../../shared/ducks/filters/filters';
import {
  getDataSelection,
  getDataSelectionResult,
  getGeometryFilter
} from '../../../shared/ducks/data-selection/selectors';
import { getViewMode } from '../../../shared/ducks/ui/ui';
import DataSelection from './DataSelection';

const mapStateToProps = (state) => {
  const { isLoading, dataset, authError, page } = getDataSelection(state);
  return ({
    isLoading,
    dataset,
    authError,
    page,
    view: getViewMode(state),
    geometryFilter: getGeometryFilter(state),
    activeFilters: getFilters(state),
    results: getDataSelectionResult(state),
    user: getUser(state),
    userScopes: getUserScopes(state)
  });
};

const mapDispatchToProps = (dispatch) => bindActionCreators({
  setPage: setDatasetPage
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(DataSelection);
