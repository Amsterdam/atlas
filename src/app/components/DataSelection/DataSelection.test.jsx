import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import DataSelection from './DataSelection';
import { DATASETS, VIEWS } from '../../../shared/ducks/data-selection/constants';

const initialState = {
  filters: {},
  user: {
    scopes: []
  },
  dataSelection: {
    isLoading: false,
    markers: [],
    geometryFilter: {
      markers: []
    },
    dataset: DATASETS.BAG,
    view: VIEWS.TABLE,
    authError: false,
    errorMessage: '',
    page: 1
  }
};

describe('DataSelection', () => {
  it('should render', () => {
    const store = configureMockStore()({ ...initialState });
    const component = shallow(<DataSelection />, { context: { store } }).dive();
    expect(component).toMatchSnapshot();
  });

  it('should render LoadingIndicator', () => {
    const state = {
      ...initialState,
      dataSelection: {
        ...initialState.dataSelection,
        isLoading: true
      }
    };
    const store = configureMockStore()({ ...state });
    const component = shallow(<DataSelection />, { context: { store } }).dive();
    expect(component).toMatchSnapshot();
  });
});
