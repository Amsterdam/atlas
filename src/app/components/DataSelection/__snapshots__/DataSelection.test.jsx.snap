// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSelection should render LoadingIndicator 1`] = `
<div
  className="c-data-selection c-dashboard__content"
>
  <div
    className="qa-data-selection-content"
  >
    <AngularWrapper
      bindings={
        Object {
          "availableFilters": undefined,
          "dataset": "bag",
          "filters": Object {},
          "isLoading": true,
          "numberOfRecords": undefined,
          "showHeader": false,
          "user": Object {
            "scopes": Array [],
          },
          "view": "TABLE",
        }
      }
      component="dpDataSelectionHeader"
      dependencies={
        Array [
          "atlas",
        ]
      }
      interpolateBindings={Object {}}
      moduleName="dpDataSelectionHeaderWrapper"
    />
    <Connect(DataSelectionActiveFilters) />
    <LoadingIndicator />
  </div>
</div>
`;

exports[`DataSelection should render as LIST 1`] = `
<div
  className="c-data-selection c-dashboard__content"
>
  <div
    className="qa-data-selection-content"
  >
    <AngularWrapper
      bindings={
        Object {
          "availableFilters": undefined,
          "dataset": "bag",
          "filters": Object {},
          "isLoading": false,
          "numberOfRecords": undefined,
          "showHeader": true,
          "user": Object {
            "scopes": Array [],
          },
          "view": "LIST",
        }
      }
      component="dpDataSelectionHeader"
      dependencies={
        Array [
          "atlas",
        ]
      }
      interpolateBindings={Object {}}
      moduleName="dpDataSelectionHeaderWrapper"
    />
    <Connect(DataSelectionActiveFilters) />
    <NoResultsForSearchType
      message="Tip: verwijder een of meer criteria"
    />
    <div
      className="u-grid qa-data-grid"
    >
      <div
        className="u-row"
      >
        <div
          className="u-col-sm--12"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`DataSelection should render as TABLE 1`] = `
<div
  className="c-data-selection c-dashboard__content"
>
  <div
    className="qa-data-selection-content"
  >
    <AngularWrapper
      bindings={
        Object {
          "availableFilters": undefined,
          "dataset": "bag",
          "filters": Object {},
          "isLoading": false,
          "numberOfRecords": undefined,
          "showHeader": true,
          "user": Object {
            "scopes": Array [],
          },
          "view": "TABLE",
        }
      }
      component="dpDataSelectionHeader"
      dependencies={
        Array [
          "atlas",
        ]
      }
      interpolateBindings={Object {}}
      moduleName="dpDataSelectionHeaderWrapper"
    />
    <Connect(DataSelectionActiveFilters) />
    <NoResultsForSearchType
      message="Tip: verwijder een of meer criteria"
    />
    <div
      className="u-grid qa-data-grid"
    >
      <div
        className="u-row"
      >
        <div
          className="u-col-sm--12"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`DataSelection should show authorization message if user has no permission 1`] = `
<div
  className="c-data-selection c-dashboard__content"
>
  <div
    className="qa-data-selection-content"
  >
    <AngularWrapper
      bindings={
        Object {
          "availableFilters": undefined,
          "dataset": "bag",
          "filters": Object {},
          "isLoading": false,
          "numberOfRecords": undefined,
          "showHeader": true,
          "user": Object {
            "scopes": Array [],
          },
          "view": "TABLE",
        }
      }
      component="dpDataSelectionHeader"
      dependencies={
        Array [
          "atlas",
        ]
      }
      interpolateBindings={Object {}}
      moduleName="dpDataSelectionHeaderWrapper"
    />
    <Connect(DataSelectionActiveFilters) />
    <NotAuthorizedMessage
      scopeError=""
    />
  </div>
</div>
`;
